### PR TITLE
#168781621 - Signup Users

### PR DESCRIPTION
**Pull request by nignanthomas**
#### What does this PR do?
Signup Users

#### Description of Task to be completed?
- Create global components Button, ToastContainer
- Create signupActions and Reducers
- Create Register view

#### How should this be manually tested?
- Clone ft-signup-168781621 branch
- run `npm i`
- run `npm run dev`
- navigate to localhost:<:port>/users
- Fill the login form
- `Register`

#### Any background context you want to provide?
N/A

#### What are the relevant pivotal tracker stories?
[#168781621](https://www.pivotaltracker.com/story/show/168781621)

#### Screenshots (if appropriate)
Screen Shot 2019-12-07 at 18 04 00
Screen Shot 2019-12-07 at 18 04 47

#### Questions:
N/A

### CODE REVIEW

**Review by alainmateso — CHANGES_REQUESTED** [review]
Great job Thomas, Kindly return a descriptive message when a user enters data that don't match the required pattern.

**Comment by ngireric123** [comment]
Great job @nignanthomas kindly try to increase test coverage

**Review by ngireric123 — CHANGES_REQUESTED** [review]
@nignanthomas kindly  increase test coverage

File: src/actions/SignupAction.js [hunk 1]
```diff
@@ -0,0 +1,54 @@
+import axios from 'axios';
+import { SIGNUP_PENDING, SIGNUP_SUCCESS, SIGNUP_FAIL } from './types';
+
+const appUrl = 'https://caret-bn-backend-staging.herokuapp.com/api/v1';
```
Comment (nakiwuge) [inline src/actions/SignupAction.js]: Configure axios with  the baseUrl

File: src/actions/SignupAction.js [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+    type: SIGNUP_SUCCESS,
+    payload: false,
+  });
+  dispatch({
+    type: SIGNUP_PENDING,
+    payload: false,
+  });
+};
+
+// export default { signupAction, resetSignup };
```
Comment (nakiwuge) [inline src/actions/SignupAction.js]: Make this file to have all auth Actions. Create one type function to be used in all the auth action functions. something like this
```
const authType = (type,payload)=>{
type,
payload
}
```
Make signUp action cleaner with something like this
```
  axios.post(`${appUrl}/users/register`, userData)
    .then((response) => {
        dispatch(authType(SIGNUP_SUCCESS,response))
    })
    .catch((error) => {
 dispatch(authType(SIGNUP_FAIL,error.reponse))
    });
};
```

File: src/actions/SignupAction.js [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+const appUrl = 'https://caret-bn-backend-staging.herokuapp.com/api/v1';
+
+export const signupAction = (userData) => (dispatch) => {
+  dispatch({
+    type: SIGNUP_PENDING,
+    payload: true,
+  });
+  axios.post(`${appUrl}/users/register`, userData)
+    .then((response) => {
+      dispatch({
```
Comment (nakiwuge) [inline src/actions/SignupAction.js]: Kindly create an `isLoading` state in the component where you call this action to cater for loading

File: src/actions/SignupAction.js [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+    payload: false,
+  });
+  dispatch({
+    type: SIGNUP_PENDING,
+    payload: false,
+  });
+};
+
+// export default { signupAction, resetSignup };
+// module.exports = { signupAction, resetSignup };
```
Comment (nakiwuge) [inline src/actions/SignupAction.js]: Remove commented code

File: src/components/features/Signup.js [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+
+  render() {
+    const { props } = this;
+    const { signup: { signupSuccess, signupFail, signupPending } } = props;
+
+    const {
+      email, username, password, confirmPassword,
+    } = this.state;
+
+    const handleChange = (e) => {
```
Comment (nakiwuge) [inline src/components/features/Signup.js]: Move this method out of the render

File: src/reducers/signupReducer.js [hunk 6]
```diff
@@ -0,0 +1,33 @@
+import { SIGNUP_PENDING, SIGNUP_SUCCESS, SIGNUP_FAIL } from '../actions/types';
+
+const initialState = {
+  signupPending: false,
+  signupFail: null,
+  signupSuccess: false,
```
Comment (nakiwuge) [inline src/reducers/signupReducer.js]: rename `signupSuccess` to `data: null`

File: src/reducers/signupReducer.js [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+        signupFail: action.payload,
+      };
+    default:
+      return {
+        ...state,
+      };
+  }
+};
+
+// export default signupReducer;
```
Comment (nakiwuge) [inline src/reducers/signupReducer.js]: remove commented code

File: src/styles/styles.scss [hunk 8]
```diff
@@ -1,21 +1,102 @@
...
+    margin: auto;
+    padding: 2rem 0;
+    background-color: $white;
+    text-align: center;
+    width: 60vw;
+    display: flex;
+    flex-direction: row;
+    justify-content: center;
+    align-items: center;
+    > div{
```
Comment (nakiwuge) [inline src/styles/styles.scss]: remove `>`

File: src/styles/styles.scss [hunk 9]
```diff
@@ -1,21 +1,102 @@
...
+			font-size: 1rem;
+			background: #073763;
+			border-color: $light;
+			color: #e2e2e2;
+			padding: 0.2rem 0.5rem;
+			cursor: pointer;
+			margin: 0.2rem 0.5rem 1rem 0;
+			transition: all 0.5s ease-in-out;
+			height: 2rem;
+			width: 150px;
```
Comment (nakiwuge) [inline src/styles/styles.scss]: Indent button

File: src/styles/styles.scss [hunk 10]
```diff
@@ -1,21 +1,102 @@
...
+    flex-direction: row;
+    justify-content: center;
+    align-items: center;
+    > div{
+        margin: 1%;
+        width: 50%;
+    }
+}
+.login-form, .signup-form{
+    > form Input{
```
Comment (nakiwuge) [inline src/styles/styles.scss]: use sass here like this 
``` form{
Input{
        width: 80%;
        padding-left: 0.5rem;
    }
} 
```

File: src/tests/SignupTest.spec.js [hunk 11]
```diff
@@ -0,0 +1,38 @@
+import React from 'react';
+import { shallow, mount } from 'enzyme';
+import { Signup } from '../components/features/Signup';
+
+const mockFunc = jest.fn();
+
```
Comment (nakiwuge) [inline src/tests/SignupTest.spec.js:10]: remove this line

File: src/tests/SignupTest.spec.js [hunk 12]
```diff
@@ -0,0 +1,38 @@
...
+    console.log(wrapper.debug());
+    
+    expect(wrapper).toHaveLength(1);
+});
+  it('Should render <Signup />', () => {
+    // expect(wrapper.exists()).toBe(true);
+  });
+
+  // it('Should render ', () => {
+  //   expect(wrapper.find('').text()).toBe('');
```
Comment (nakiwuge) [inline src/tests/SignupTest.spec.js]: remove this commented code

File: src/tests/actionsTests/SignupActionsTest.spec.js [hunk 13]
```diff
@@ -0,0 +1,145 @@
+import { SIGNUP_PENDING, SIGNUP_SUCCESS, SIGNUP_FAIL } from '../../actions/types';
+import { signupAction, resetSignup } from '../../actions/SignupAction';
+// import store from '../../reduxStore';
```
Comment (nakiwuge) [inline src/tests/actionsTests/SignupActionsTest.spec.js]: remove this commented code

File: src/tests/actionsTests/SignupActionsTest.spec.js [hunk 14]
```diff
@@ -0,0 +1,145 @@
...
+import configureMockStore from 'redux-mock-store';
+import thunk from 'redux-thunk';
+
```
Comment (nakiwuge) [inline src/tests/actionsTests/SignupActionsTest.spec.js:11]: remove this line

**Review by ngireric123 — APPROVED** [review]
LGTM

**Review by alainmateso — APPROVED** [review]

**Pull request merged**